File: src/Sample/Plumix.Sample/Demos/General/RouterDemoPage.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/general/router_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class RouterDemoPage : StatefulWidget
{
    public override State CreateState() => new RouterDemoPageState();
}

internal sealed class RouterDemoPageState : State
{
    private readonly DemoRouteInformationProvider _provider =
        new(new RouteInformation(new Uri("/home", UriKind.Relative)));

    private readonly DemoRouterDelegate _routerDelegate = new();
    private readonly RootBackButtonDispatcher _backButtonDispatcher = new();
    private readonly List<string> _reports = [];

    public override void InitState()
    {
        base.InitState();
        _provider.Reported = information => SetState(() => _reports.Add(information.Uri.ToString()));
    }

    public override void Dispose()
    {
        _provider.Dispose();
        _routerDelegate.Dispose();
        base.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new Text("Router demo", fontSize: 20, color: Colors.Black),
                new Text(
                    "The provider publishes a location, the parser turns it into a configuration and the "
                    + "delegate builds the page. Popping goes through the back-button dispatcher.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildAction("Home", () => Go("/home"), Color.Parse("#FFE8F5E9")),
                        BuildAction("Details", () => Go("/details"), Color.Parse("#FFE3F2FD")),
                        BuildAction("Settings", () => Go("/settings"), Color.Parse("#FFFFF3E0")),
                        BuildAction("Back", HandleBack, Color.Parse("#FFFCE4EC")),
                    ]),
                new Text($"reported: {string.Join(", ", _reports)}", fontSize: 12),
                new SizedBox(
                    height: 180,
                    child: new ColoredBox(
                        color: Color.Parse("#FFFAFAFA"),
                        child: new Router<DemoRouteConfiguration>(
                            routerDelegate: _routerDelegate,
                            routeInformationProvider: _provider,
                            routeInformationParser: new DemoRouteInformationParser(),
                            backButtonDispatcher: _backButtonDispatcher))),
            ]);
    }

    private void Go(string location)
    {
        _provider.SetValue(new RouteInformation(new Uri(location, UriKind.Relative)));
    }

    private void HandleBack()
    {
        _ = _backButtonDispatcher.InvokeCallback(Task.FromResult(false));
    }

    private static Widget BuildAction(string label, Action onTap, Color background)
    {
        return new CounterTapButton(
            label: label,
            onTap: onTap,
            background: background,
            foreground: Colors.Black,
            fontSize: 12,
            padding: new Thickness(10, 8));
    }
}

/// <summary>The parsed configuration the demo delegate renders.</summary>
internal sealed record DemoRouteConfiguration(string Path, bool ShowDetail);

internal sealed class DemoRouteInformationParser : RouteInformationParser<DemoRouteConfiguration>
{
    public override Task<DemoRouteConfiguration> ParseRouteInformation(RouteInformation routeInformation)
    {
        string path = routeInformation.Uri.ToString();
        return Task.FromResult(new DemoRouteConfiguration(path, path == "/details"));
    }

    public override RouteInformation? RestoreRouteInformation(DemoRouteConfiguration configuration)
    {
        return new RouteInformation(new Uri(configuration.Path, UriKind.Relative));
    }
}

internal sealed class DemoRouterDelegate : RouterDelegate<DemoRouteConfiguration>
{
    private DemoRouteConfiguration _configuration = new("/home", ShowDetail: false);

    public override DemoRouteConfiguration? CurrentConfiguration => _configuration;

    public override Task SetNewRoutePath(DemoRouteConfiguration configuration)
    {
        _configuration = configuration;
        return Task.CompletedTask;
    }

    public override Task<bool> PopRoute()
    {
        if (_configuration.Path == "/home")
        {
            return Task.FromResult(false);
        }

        _configuration = new DemoRouteConfiguration("/home", ShowDetail: false);
        NotifyListeners();
        return Task.FromResult(true);
    }

    public override Widget Build(BuildContext context)
    {
        return new BackButtonListener(
            onBackButtonPressed: PopRoute,
            child: new Center(
                child: new Column(
                    mainAxisSize: MainAxisSize.Min,
                    spacing: 8,
                    children:
                    [
                        new Text(_configuration.Path, fontSize: 18, color: Colors.Black),
                        new Text(
                            _configuration.ShowDetail ? "detail page" : "top level page",
                            fontSize: 12,
                            color: Colors.DimGray),
                    ])));
    }
}

internal sealed class DemoRouteInformationProvider : RouteInformationProvider
{
    private RouteInformation _value;

    public DemoRouteInformationProvider(RouteInformation value)
    {
        _value = value;
    }

    public Action<RouteInformation>? Reported { get; set; }

    public override RouteInformation Value => _value;

    public void SetValue(RouteInformation value)
    {
        _value = value;
        NotifyListeners();
    }

    public override void RouterReportsNewRouteInformation(
        RouteInformation routeInformation,
        RouteInformationReportingType type = RouteInformationReportingType.None)
    {
        _value = routeInformation;
        Reported?.Invoke(routeInformation);
    }
}

File: dart_sample/lib/demos/general/router_demo_page.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

/// C# parity source: src/Sample/Plumix.Sample/Demos/General/RouterDemoPage.cs
class RouterDemoPage extends StatefulWidget {
  const RouterDemoPage({super.key});

  @override
  State<RouterDemoPage> createState() => _RouterDemoPageState();
}

class _RouterDemoPageState extends State<RouterDemoPage> {
  final _DemoRouteInformationProvider _provider = _DemoRouteInformationProvider(
    RouteInformation(uri: Uri.parse('/home')),
  );
  final _DemoRouterDelegate _routerDelegate = _DemoRouterDelegate();
  final RootBackButtonDispatcher _backButtonDispatcher = RootBackButtonDispatcher();
  final List<String> _reports = <String>[];

  @override
  void initState() {
    super.initState();
    _provider.reported = (RouteInformation information) {
      setState(() => _reports.add(information.uri.toString()));
    };
  }

  @override
  void dispose() {
    _provider.dispose();
    _routerDelegate.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        const Text(
          'Router demo',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'The provider publishes a location, the parser turns it into a configuration and the '
          'delegate builds the page. Popping goes through the back-button dispatcher.',
          style: TextStyle(fontSize: 14, color: Colors.grey),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildAction('Home', () => _go('/home'), const Color(0xFFE8F5E9)),
            _buildAction('Details', () => _go('/details'), const Color(0xFFE3F2FD)),
            _buildAction('Settings', () => _go('/settings'), const Color(0xFFFFF3E0)),
            _buildAction('Back', _handleBack, const Color(0xFFFCE4EC)),
          ],
        ),
        Text(
          'reported: ${_reports.join(', ')}',
          style: const TextStyle(fontSize: 12),
        ),
        SizedBox(
          height: 180,
          child: ColoredBox(
            color: const Color(0xFFFAFAFA),
            child: Router<_DemoRouteConfiguration>(
              routerDelegate: _routerDelegate,
              routeInformationProvider: _provider,
              routeInformationParser: _DemoRouteInformationParser(),
              backButtonDispatcher: _backButtonDispatcher,
            ),
          ),
        ),
      ],
    );
  }

  void _go(String location) {
    _provider.setValue(RouteInformation(uri: Uri.parse(location)));
  }

  void _handleBack() {
    _backButtonDispatcher.invokeCallback(SynchronousFuture<bool>(false));
  }

  static Widget _buildAction(String label, VoidCallback onTap, Color background) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
        color: background,
        child: Text(
          label,
          style: const TextStyle(fontSize: 12, color: Colors.black),
        ),
      ),
    );
  }
}

/// The parsed configuration the demo delegate renders.
@immutable
class _DemoRouteConfiguration {
  const _DemoRouteConfiguration(this.path, {required this.showDetail});

  final String path;
  final bool showDetail;
}

class _DemoRouteInformationParser extends RouteInformationParser<_DemoRouteConfiguration> {
  @override
  Future<_DemoRouteConfiguration> parseRouteInformation(RouteInformation routeInformation) {
    final String path = routeInformation.uri.toString();
    return SynchronousFuture<_DemoRouteConfiguration>(
      _DemoRouteConfiguration(path, showDetail: path == '/details'),
    );
  }

  @override
  RouteInformation? restoreRouteInformation(_DemoRouteConfiguration configuration) {
    return RouteInformation(uri: Uri.parse(configuration.path));
  }
}

class _DemoRouterDelegate extends RouterDelegate<_DemoRouteConfiguration> with ChangeNotifier {
  _DemoRouteConfiguration _configuration = const _DemoRouteConfiguration('/home', showDetail: false);

  @override
  _DemoRouteConfiguration? get currentConfiguration => _configuration;

  @override
  Future<void> setNewRoutePath(_DemoRouteConfiguration configuration) {
    _configuration = configuration;
    return SynchronousFuture<void>(null);
  }

  @override
  Future<bool> popRoute() {
    if (_configuration.path == '/home') {
      return SynchronousFuture<bool>(false);
    }

    _configuration = const _DemoRouteConfiguration('/home', showDetail: false);
    notifyListeners();
    return SynchronousFuture<bool>(true);
  }

  @override
  Widget build(BuildContext context) {
    return BackButtonListener(
      onBackButtonPressed: popRoute,
      child: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          spacing: 8,
          children: <Widget>[
            Text(
              _configuration.path,
              style: const TextStyle(fontSize: 18, color: Colors.black),
            ),
            Text(
              _configuration.showDetail ? 'detail page' : 'top level page',
              style: const TextStyle(fontSize: 12, color: Colors.grey),
            ),
          ],
        ),
      ),
    );
  }
}

class _DemoRouteInformationProvider extends RouteInformationProvider with ChangeNotifier {
  _DemoRouteInformationProvider(this._value);

  RouteInformation _value;

  ValueChanged<RouteInformation>? reported;

  @override
  RouteInformation get value => _value;

  void setValue(RouteInformation value) {
    _value = value;
    notifyListeners();
  }

  @override
  void routerReportsNewRouteInformation(
    RouteInformation routeInformation, {
    RouteInformationReportingType type = RouteInformationReportingType.none,
  }) {
    _value = routeInformation;
    reported?.call(routeInformation);
  }
}
